File: backend_py/core/bot.py
import asyncio
from typing import Dict, Any, Optional, List
from datetime import datetime
from loguru import logger
from pydantic import BaseModel

from .exchange import ExchangeManager, MarketData

class BotConfig(BaseModel):
    symbol: str
    timeframe: str
    trading_enabled: bool
    strategy: str  # 'ema_ml' | 'price_action' | 'macd_rsi' | 'bollinger'

class Position(BaseModel):
    symbol: str
    side: str  # 'long' | 'short'
    entry_price: float
    current_price: float
    amount: float
    profit: float
    profit_percent: float
    stop_loss: float
    take_profit: float
    open_time: str

class TradingBot:
    def __init__(self, exchange: ExchangeManager, config: Dict[str, Any], ml_service_url: str = 'http://localhost:5000'):
        self.exchange = exchange
        self.config = BotConfig(**config)
        self.start_time = datetime.now().timestamp() * 1000

        # Импорт здесь для избежания циклических импортов
        from .risk_manager import RiskManager
        from ..services.ml_service import MLService
        from ..strategies.ema_strategy import EMAStrategy

        # Инициализация компонентов
        self.risk_manager = RiskManager({
            'max_position_percent': 10,
            'max_daily_loss_percent': 5,
            'max_drawdown_percent': 10,
            'min_order_size': 10,
            'max_order_size': 1000
        })

        self.ml_service = MLService(ml_service_url)
        self.strategy = EMAStrategy(exchange, config['symbol'], config['timeframe'])

        self.is_running = False
        self.trading_enabled = config.get('trading_enabled', False)
        self.current_position: Optional[Position] = None
        self.trading_loop_task: Optional[asyncio.Task] = None

    async def start(self):
        if self.is_running:
            return

        self.is_running = True
        logger.info(f'🚀 Trading Bot started for {self.config.symbol} with strategy {self.config.strategy}')

        # Запускаем торговый цикл
        self.trading_loop_task = asyncio.create_task(self._trading_loop())

    async def stop(self):
        self.is_running = False
        if self.trading_loop_task:
            self.trading_loop_task.cancel()
            try:
                await self.trading_loop_task
            except asyncio.CancelledError:
                pass
        logger.info('🛑 Trading Bot stopped')

    async def _trading_loop(self):
        while self.is_running:
            try:
                if self.trading_enabled:
                    await self._check_signals()
                    await self._manage_position()
                await asyncio.sleep(60)  # Проверяем каждую минуту
            except Exception as e:
                logger.error(f'Trading loop error: {e}')
                await asyncio.sleep(60)

    async def _check_signals(self):
        try:
            # Получаем сигнал от стратегии
            signal = await self.strategy.get_signal()

            if signal and not self.current_position:
                # Открываем позицию
                await self._open_position(signal)
            elif signal == 'close' and self.current_position:
                # Закрываем позицию
                await self._close_position()
        except Exception as e:
            logger.error(f'Failed to check signals: {e}')

    async def _open_position(self, signal: str):
        try:
            market_data = await self.exchange.get_market_data(self.config.symbol)
            if not market_data:
                return

            # Рассчитываем размер позиции через Risk Manager
            balance = await self.exchange.get_balance('USDT')
            position_size = self.risk_manager.calculate_position_size(balance.available, market_data.price)

            if position_size < self.risk_manager.config.min_order_size:
                logger.warning(f'Position size too small: {position_size}')
                return

            # Размещаем ордер
            side = 'buy' if signal == 'buy' else 'sell'
            order = await self.exchange.place_order(
                self.config.symbol,
                side,
                position_size / market_data.price,  # Количество в базовой валюте
                market_data.price
            )

            # Создаем позицию
            self.current_position = Position(
                symbol=self.config.symbol,
                side='long' if side == 'buy' else 'short',
                entry_price=market_data.price,
                current_price=market_data.price,
                amount=position_size / market_data.price,
                profit=0,
                profit_percent=0,
                stop_loss=self.risk_manager.calculate_stop_loss(market_data.price, 'long' if side == 'buy' else 'short'),
                take_profit=self.risk_manager.calculate_take_profit(market_data.price, 'long' if side == 'buy' else 'short'),
                open_time=datetime.now().isoformat()
            )

            logger.info(f'Position opened: {self.current_position}')
        except Exception as e:
            logger.error(f'Failed to open position: {e}')

    async def _close_position(self):
        try:
            if not self.current_position:
                return

            market_data = await self.exchange.get_market_data(self.config.symbol)
            if not market_data:
                return

            # Закрываем позицию
            side = 'sell' if self.current_position.side == 'long' else 'buy'
            order = await self.exchange.place_order(
                self.config.symbol,
                side,
                self.current_position.amount,
                market_data.price
            )

            # Обновляем позицию
            self.current_position.current_price = market_data.price
            self.current_position.profit = (market_data.price - self.current_position.entry_price) * self.current_position.amount
            self.current_position.profit_percent = (self.current_position.profit / (self.current_position.entry_price * self.current_position.amount)) * 100

            logger.info(f'Position closed: {self.current_position}')
            self.current_position = None
        except Exception as e:
            logger.error(f'Failed to close position: {e}')

    async def _manage_position(self):
        if not self.current_position:
            return

        try:
            market_data = await self.exchange.get_market_data(self.config.symbol)
            if not market_data:
                return

            # Проверяем стоп-лосс и тейк-профит
            if self.current_position.side == 'long':
                if market_data.price <= self.current_position.stop_loss or market_data.price >= self.current_position.take_profit:
                    await self._close_position()
            else:  # short
                if market_data.price >= self.current_position.stop_loss or market_data.price <= self.current_position.take_profit:
                    await self._close_position()
        except Exception as e:
            logger.error(f'Failed to manage position: {e}')

    async def get_status(self) -> Dict[str, Any]:
        return {
            'is_running': self.is_running,
            'trading_enabled': self.trading_enabled,
            'symbol': self.config.symbol,
            'strategy': self.config.strategy,
            'current_position': self.current_position.dict() if self.current_position else None,
            'uptime': datetime.now().timestamp() * 1000 - self.start_time
        }

    async def get_market_data(self) -> Optional[MarketData]:
        return await self.exchange.get_market_data(self.config.symbol)

    def enable_trading(self):
        self.trading_enabled = True
        logger.info('Trading enabled')

    def disable_trading(self):
        self.trading_enabled = False
        logger.info('Trading disabled')

File: backend/src/core/bot.ts
import { ExchangeManager } from './exchange';
import { RiskManager } from './riskManager';
import { MLService } from '../services/mlService';
import { EMAStrategy } from '../strategies/EMAStrategy';
import { logger } from '../utils/logger';

export interface BotConfig {
  symbol: string;
  timeframe: string;
  tradingEnabled: boolean;
  strategy: 'ema_ml' | 'price_action' | 'macd_rsi' | 'bollinger';
}

export interface Position {
  symbol: string;
  side: 'long' | 'short';
  entryPrice: number;
  currentPrice: number;
  amount: number;
  profit: number;
  profitPercent: number;
  stopLoss: number;
  takeProfit: number;
  openTime: string;
}

export class TradingBot {
  private exchange: ExchangeManager;
  private riskManager: RiskManager;
  private mlService: MLService;
  private strategy: EMAStrategy;
  private config: BotConfig;
  private isRunning: boolean = false;
  private tradingEnabled: boolean = false;
  private currentPosition: Position | null = null;
  private startTime: number;
  private tradingLoopInterval: NodeJS.Timeout | null = null;

  constructor(
    exchange: ExchangeManager,
    config: BotConfig,
    mlServiceURL?: string
  ) {
    this.exchange = exchange;
    this.config = config;
    this.startTime = Date.now();

    // Инициализация Risk Manager
    this.riskManager = new RiskManager({
      maxPositionPercent: 10,
      stopLossPercent: 2,
      takeProfitPercent: 5,
      maxDailyTrades: 10,
      minTradeInterval: 300
    });

    // Инициализация ML Service
    this.mlService = new MLService({
      baseURL: mlServiceURL || process.env.ML_SERVICE_URL || 'http://localhost:5000'
    });

    // Инициализация стратегии
    this.strategy = new EMAStrategy({
      fastPeriod: 9,
      slowPeriod: 21,
      threshold: 0.25
    });

    logger.info('Trading Bot initialized');
    logger.info(`Symbol: ${config.symbol}`);
    logger.info(`Strategy: ${config.strategy}`);
  }  /**
   * Запустить бота
   */
  async start(): Promise<void> {
    if (this.isRunning) {
      logger.warn('Bot is already running');
      return;
    }

    try {
      // Проверяем подключение к бирже
      const connected = await this.exchange.connect();
      if (!connected) {
        throw new Error('Failed to connect to exchange');
      }

      // Проверяем ML Service (необязательно)
      await this.mlService.checkHealth();

      this.isRunning = true;
      logger.info('🚀 Trading Bot started');
      
      // Запускаем основной торговый цикл
      this.runTradingLoop();
    } catch (error) {
      logger.error('Failed to start bot:', error);
      throw error;
    }
  }

  /**
   * Остановить бота
   */
  stop(): void {
    if (!this.isRunning) {
      logger.warn('Bot is not running');
      return;
    }

    // Останавливаем торговый цикл
    if (this.tradingLoopInterval) {
      clearInterval(this.tradingLoopInterval);
      this.tradingLoopInterval = null;
    }

    this.isRunning = false;
    logger.info('🛑 Trading Bot stopped');
  }

  /**
   * Включить торговлю
   */
  enableTrading(): void {
    this.tradingEnabled = true;
    logger.info('✅ Trading enabled');
  }

  /**
   * Отключить торговлю
   */
  disableTrading(): void {
    this.tradingEnabled = false;
    logger.info('⚠️ Trading disabled');
  }

  /**
   * Получить статус бота
   */
  async getStatus(): Promise<any> {
    try {
      const balance = await this.exchange.getBalance('USDT');
      const uptime = Math.floor((Date.now() - this.startTime) / 1000);

      return {
        isRunning: this.isRunning,
        tradingEnabled: this.tradingEnabled,
        balance: {
          total: balance.total,
          available: balance.available,
          used: balance.used,
          currency: balance.currency
        },
        positions: {
          current: this.currentPosition,
          total: this.currentPosition ? 1 : 0,
          profit: this.currentPosition?.profit || 0
        },
        uptime,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      logger.error('Failed to get bot status:', error);
      throw error;
    }
  }

  /**
   * Получить рыночные данные
   */
  async getMarketData(): Promise<any> {
    try {
      return await this.exchange.getMarketData(this.config.symbol);
    } catch (error) {
      logger.error('Failed to get market data:', error);
      throw error;
    }
  }

  /**
   * Получить текущую позицию
   */
  getCurrentPosition(): Position | null {
    return this.currentPosition;
  }

  /**
   * Проверить, запущен ли бот
   */
  isActive(): boolean {
    return this.isRunning;
  }

  /**
   * Проверить, включена ли торговля
   */
  isTradingEnabled(): boolean {
    return this.tradingEnabled;
  }

  /**
   * Обновить конфигурацию
   */
  updateConfig(config: Partial<BotConfig>): void {
    this.config = { ...this.config, ...config };
    logger.info('Bot configuration updated:', config);
  }

  /**
   * Получить uptime в секундах
   */
  getUptime(): number {
    return Math.floor((Date.now() - this.startTime) / 1000);
  }

  /**
   * Основной торговый цикл
   */
  private runTradingLoop(): void {
    // Запускаем цикл каждые 30 секунд
    this.tradingLoopInterval = setInterval(async () => {
      if (!this.isRunning) {
        return;
      }

      try {
        await this.executeTradingCycle();
      } catch (error) {
        logger.error('Error in trading cycle:', error);
      }
    }, 30000); // 30 секунд

    logger.info('Trading loop started (30s interval)');
  }

  /**
   * Выполнить один цикл торговли
   */
  private async executeTradingCycle(): Promise<void> {
    try {
      // Получаем текущую цену и OHLCV данные
      const ticker = await this.exchange.getTicker(this.config.symbol);
      const ohlcv = await this.exchange.getOHLCV(
        this.config.symbol,
        this.config.timeframe,
        100
      );

      const currentPrice = ticker.last;

      logger.debug(`Trading cycle: ${this.config.symbol} @ ${currentPrice}`);

      // Если есть открытая позиция, проверяем нужно ли её закрыть
      if (this.currentPosition) {
        await this.checkPositionExit(currentPrice);
        return; // Не открываем новую позицию, если уже есть открытая
      }

      // Если торговля не включена, не открываем позиции
      if (!this.tradingEnabled) {
        return;
      }

      // Проверяем можно ли открыть новую сделку (risk management)
      const canTrade = this.riskManager.canOpenTrade();
      if (!canTrade.allowed) {
        logger.debug(`Cannot trade: ${canTrade.reason}`);
        return;
      }

      // Анализируем рынок через EMA стратегию
      const emaSignal = this.strategy.analyze(ohlcv, currentPrice);
      logger.debug(`EMA Signal: ${emaSignal.action} (confidence: ${(emaSignal.confidence * 100).toFixed(1)}%)`);

      // Если ML Service доступен, получаем ML предсказание
      let mlSignal = null;
      if (this.mlService.isServiceAvailable()) {
        const features = this.mlService.prepareFeatures(ohlcv);
        if (features.length > 0) {
          mlSignal = await this.mlService.predict(features, ohlcv);
          logger.debug(`ML Signal: ${mlSignal.signal} (confidence: ${(mlSignal.confidence * 100).toFixed(1)}%)`);
        }
      }

      // Комбинируем сигналы
      const finalSignal = this.combineSignals(emaSignal, mlSignal);

      // Открываем позицию если сигнал сильный
      if (finalSignal.action === 'BUY' && finalSignal.confidence > 0.6) {
        await this.openPosition('buy', currentPrice, finalSignal.reason);
      } else if (finalSignal.action === 'SELL' && finalSignal.confidence > 0.6) {
        await this.openPosition('sell', currentPrice, finalSignal.reason);
      }

    } catch (error) {
      logger.error('Error in trading cycle:', error);
    }
  }

  /**
   * Комбинировать EMA и ML сигналы
   */
  private combineSignals(emaSignal: any, mlSignal: any): any {
    if (!mlSignal) {
      return emaSignal; // Только EMA
    }

    // Если оба сигнала согласны
    if (emaSignal.action === mlSignal.signal) {
      return {
        action: emaSignal.action,
        confidence: (emaSignal.confidence + mlSignal.confidence) / 2,
        reason: `EMA + ML agree: ${emaSignal.action}`
      };
    }

    // Если сигналы противоречат, берем более уверенный
    if (emaSignal.confidence > mlSignal.confidence) {
      return {
        action: emaSignal.action,
        confidence: emaSignal.confidence * 0.7, // Снижаем уверенность
        reason: `EMA stronger: ${emaSignal.action}`
      };
    } else {
      return {
        action: mlSignal.signal,
        confidence: mlSignal.confidence * 0.7,
        reason: `ML stronger: ${mlSignal.signal}`
      };
    }
  }

  /**
   * Открыть позицию
   */
  private async openPosition(
    side: 'buy' | 'sell',
    currentPrice: number,
    reason: string
  ): Promise<void> {
    try {
      // Получаем баланс
      const balance = await this.exchange.getBalance('USDT');

      // Рассчитываем размер позиции через Risk Manager
      const tradeSize = this.riskManager.calculatePositionSize(
        balance.available,
        currentPrice,
        side
      );

      logger.info(`🔷 Opening ${side.toUpperCase()} position:`);
      logger.info(`   Amount: ${tradeSize.amountInCurrency.toFixed(6)} ${this.config.symbol.split('/')[0]}`);
      logger.info(`   Size: $${tradeSize.amountInUSDT.toFixed(2)}`);
      logger.info(`   Stop Loss: $${tradeSize.stopLoss.toFixed(2)}`);
      logger.info(`   Take Profit: $${tradeSize.takeProfit.toFixed(2)}`);
      logger.info(`   Reason: ${reason}`);

      // ВАЖНО: В production здесь создается реальный ордер
      // const order = await this.exchange.createMarketOrder(
      //   this.config.symbol,
      //   side,
      //   tradeSize.amountInCurrency
      // );

      // Сохраняем позицию
      this.currentPosition = {
        symbol: this.config.symbol,
        side: side === 'buy' ? 'long' : 'short',
        entryPrice: currentPrice,
        currentPrice: currentPrice,
        amount: tradeSize.amountInCurrency,
        profit: 0,
        profitPercent: 0,
        stopLoss: tradeSize.stopLoss,
        takeProfit: tradeSize.takeProfit,
        openTime: new Date().toISOString()
      };

      // Регистрируем сделку в Risk Manager
      this.riskManager.registerTrade();

      logger.info('✅ Position opened successfully');

    } catch (error) {
      logger.error('Failed to open position:', error);
    }
  }

  /**
   * Проверить нужно ли закрыть позицию
   */
  private async checkPositionExit(currentPrice: number): Promise<void> {
    if (!this.currentPosition) return;

    // Обновляем текущую цену и прибыль
    const { profit, profitPercent } = this.riskManager.calculateProfitLoss(
      this.currentPosition.entryPrice,
      currentPrice,
      this.currentPosition.amount,
      this.currentPosition.side
    );

    this.currentPosition.currentPrice = currentPrice;
    this.currentPosition.profit = profit;
    this.currentPosition.profitPercent = profitPercent;

    // Проверяем Stop Loss / Take Profit
    const shouldClose = this.riskManager.shouldClosePosition(
      this.currentPosition.entryPrice,
      currentPrice,
      this.currentPosition.side,
      this.currentPosition.stopLoss,
      this.currentPosition.takeProfit
    );

    if (shouldClose.shouldClose) {
      await this.closePosition(shouldClose.reason || 'Unknown reason');
    }
  }

  /**
   * Закрыть позицию
   */
  private async closePosition(reason: string): Promise<void> {
    if (!this.currentPosition) return;

    try {
      logger.info(`🔶 Closing ${this.currentPosition.side.toUpperCase()} position:`);
      logger.info(`   Profit: $${this.currentPosition.profit.toFixed(2)} (${this.currentPosition.profitPercent.toFixed(2)}%)`);
      logger.info(`   Reason: ${reason}`);

      // ВАЖНО: В production здесь создается ордер на закрытие
      // const side = this.currentPosition.side === 'long' ? 'sell' : 'buy';
      // await this.exchange.createMarketOrder(
      //   this.config.symbol,
      //   side,
      //   this.currentPosition.amount
      // );

      logger.info('✅ Position closed successfully');

      // Очищаем позицию
      this.currentPosition = null;

    } catch (error) {
      logger.error('Failed to close position:', error);
    }
  }
}
